### ISSUE TITLE
Laisser ErrorMiddleware s'occuper des erreurs HTTP dans les contrôleurs

### ISSUE THREAD

**Issue by mxmeunier** [betagouv/api-subventions-asso#761]
CECI EST UNE EPIQUE

Faire en sorte que les services appelés dans les contrôleurs retournes une erreur contenant un message et un code.

ErrorMiddleware se charge déjà de renvoyer un réponse formatée avec définissant le code erreur de la requête et en insérant le message.

TODO: Créer un ticket pour chaque contrôleur

**Comment by alice-telescoop** [comment]
Comment gère-t-on les tests avec ce middleware ? Avant ces retours d'erreurs étaient testés dans les tests unitaires du controller mais le middleware ne fait pas effet dans ces tests. On ne les teste donc plus là, parce que le middleware est testé à part ? @VSerain @mxmeunier

**Comment by VSerain** [comment]
> Comment gère-t-on les tests avec ce middleware ? Avant ces retours d'erreurs étaient testés dans les tests unitaires du controller mais le middleware ne fait pas effet dans ces tests. On ne les teste donc plus là, parce que le middleware est testé à part ? @VSerain @mxmeunier

Hello, normalement, tu peux tester que ton contrôleur http throw bien une erreur ! 
Je sais pas si ma réponse correspond à ta question ? @alice-telescoop

**Comment by alice-telescoop** [comment]
> Hello, normalement, tu peux tester que ton contrôleur http throw bien une erreur ! Je sais pas si ma réponse correspond à ta question ? @alice-telescoop

Avec [.toThrowError()](https://jestjs.io/docs/expect#tothrowerror) ? On a un exemple d'utilisation ?

**Comment by VSerain** [comment]
> > Hello, normalement, tu peux tester que ton contrôleur http throw bien une erreur ! Je sais pas si ma réponse correspond à ta question ? @alice-telescoop
> 
> Avec [.toThrowError()](https://jestjs.io/docs/expect#tothrowerror) ? On a un exemple d'utilisation ?

Oui avec sois toThrowError si c'est pas async sois expect().rejects.toThrowError

Par exemple dans : 
- `packages/api/src/authentication/authentication.test.ts`

Mais c'est vrai que j'ai l'impression qu'on n'a jamais testé ces cas la ! C'est pas ouf, ça sera à rajouter pour chaque sous tickets de cette epic

**Comment by mxmeunier** [comment]
Selon moi c'est à tester en test d'intégration avec supertest. On fait un simple appel HTTP et on s'assure qu'on nous retourne bien une erreur

**Comment by alice-telescoop** [comment]
J'aime bien l'idée de quand même tester ça en unitaire parce que c'est vraiment le taff du controller de faire qqchose dans les cas qui doivent générer une erreur. Ça n'empêche pas les tests d'intégration dessus mais pour moi les tests unitaires devraient couvrir ça. Mais du coup on sait comment

**Comment by mxmeunier** [comment]
Je ne suis pas d'accord te tester unitairement le ErrorMiddleware pour chaque contrôleur. Ce middleware est utilisé sur tous les contrôleurs donc il faut :

- tester unitairement la function errorHandler sur ErrorMiddleware (dans un ErrorMiddleware.test.ts)
- tester de bout en bout l'appel aux contrôleurs pour déclencher l'erreur (= test d'intégration)

**Comment by alice-telescoop** [comment]
Je suis d'accord pour ne pas tester le middleware à chaque fois (désolée je n'ai pas été claire)
Par contre le middleware est déclenchée si le controller lance une exception et je pense que ce déclenchement d'erreur devrait être testé unitairement

**Comment by mxmeunier** [comment]
Ce ne serait donc pas un test unitaire mais d'intégration puisque ça prend en compte l'exécution des middleware Express pour intercepter l'erreur. D'où l'intérêt de mettre ça dans un fichier `.spec.ts`

**Comment by alice-telescoop** [comment]
Non, justement on ne teste pas l'interception de l'erreur mais juste le fait qu'elle est envoyée

**Comment by mxmeunier** [comment]
Je ne comprend pas. C'est soit déjà fait en testant unitairement la méthode de service appelée soit pas besoin de test puisque l'utilisation du ErrorMiddleware supprime le try/catch donc n'importe quel throw d'erreur depuis le service déclencherait le mécanisme

**Comment by alice-telescoop** [comment]
Mais le controller peut envoyer des erreurs aussi, non ?

**Comment by mxmeunier** [comment]
Si on utilise le ErrorMiddleware alors les contrôleurs ne renvoient qu'un appel aux services et c'est aux services de déterminer l'erreur à remonter (sous la forme d'une HttpError qui prend un code et un message) et si aucune erreur n'est déterminée alors le ErrorMIddleware renvoie un 500 Internal Server Error

**Comment by VSerain** [comment]
On en reparle au point de toute à l'heure @mxmeunier @alice-telescoop
